js/store.js

```diff
@@ -112,51 +216,139 @@ const store = {
+  },
+
+  // check if third party is on the whitelist (owned by the first party)
+  // returns true if it is and false otherwise
+  onWhiteList(firstPartyFromRequest, thirdPartyFromRequest) {
+    if (thirdPartyFromRequest && this.firstPartyWhiteList) {
+      const hostnameVariantsFirstParty
+        = this.getHostnameVariants(firstPartyFromRequest);
+      for (let i = 0; i < hostnameVariantsFirstParty.length; i++) {
+        if (this.firstPartyWhiteList
+          .hasOwnProperty(hostnameVariantsFirstParty[i])) {
+          // first party is in the whitelist
+          const index = this.firstPartyWhiteList[hostnameVariantsFirstParty[i]];
+          const hostnameVariantsThirdParty
+            = this.getHostnameVariants(thirdPartyFromRequest);
+          for (let j = 0; j < hostnameVariantsThirdParty.length; j++) {
+            if (this.thirdPartyWhiteList[index]
+              .includes(hostnameVariantsThirdParty[j])) {
+              return true;
+            }
+          }
+          return false;
+        }
+      }
+    }
+    return false;
   },
 
   setFirstParty(hostname, data) {
     if (!hostname) {
       throw new Error('setFirstParty requires a valid hostname argument');
     }
+    const isNewWebsite = this.isNewWebsite(hostname);
+
     this.setWebsite(hostname, data);
+
+    if (isNewWebsite) {
+      this.updateChild(this.outputWebsite(hostname, this._websites[hostname]));
+    }
   },
 
+  // @todo use data from capture to update UI/filter graph
+  // eslint-disable-next-line no-unused-vars
   setThirdParty(origin, target, data) {
-    let newThirdParty = false;
+    let isNewWebsite = false;
     if (!origin) {
       throw new Error('setThirdParty requires a valid origin argument');
     }
 
-    const firstParty = this.getWebsite(origin);
-    const thirdParty = this.getWebsite(target);
+    let firstParty = this.getWebsite(origin);
+    let thirdParty = this.getWebsite(target);
 
     if (!('thirdPartyHostnames' in firstParty)) {
       firstParty['thirdPartyHostnames'] = [];
     }
     if (!firstParty['thirdPartyHostnames'].includes(target)) {
-      firstParty['thirdPartyHostnames'].push(target);
-      newThirdParty = true;
+      // third party is not currently linked to first party
+      ({ firstParty, thirdParty, isNewWebsite }
+        = this.whiteListCheck(origin, target, firstParty, thirdParty));
     }
-    if (!('firstPartyHostnames' in firstParty)) {
+    if (!('firstPartyHostnames' in thirdParty)) {
       thirdParty['firstPartyHostnames'] = [];
     }
     if (!thirdParty['firstPartyHostnames'].includes(origin)) {
       thirdParty['firstPartyHostnames'].push(origin);
-      newThirdParty = true;
     }
 
-    this.setWebsite(origin, firstParty);
+    this.setFirstParty(origin, firstParty);
     this.setWebsite(target, thirdParty);
 
-    if (newThirdParty) {
-      this.updateChild(this.outputWebsite(target, data, origin));
+    if (isNewWebsite && thirdParty['isVisible']) {
```

Comment:
I can see for sure that we need to change a few things in your code:
1. Add the code below before the nested block of `if` statements in `setThirdParty`; since the first party won't have this key initialized for the very first third party request it makes.
```javascript
    if (!('thirdPartyHostnames' in firstParty)) {
      firstParty['thirdPartyHostnames'] = [];
    }
```
2. You also never initialize `thirdParty['isVisible'] to false if it is whitelisted. I added an `else` statement for this case.
3. You call `getWebsite` for `firstParty` inside `setThirdParty`, then you call it again in `addFirstPartyRelationship`. Since we're accessing `firstParty['thirdPartyHostnames']` in the former as the top level `if` statement (in the set of nested `ifs`), I have passed `firstParty` into the latter instead of re-getting it there and repeating the same logic to build it up in the helper method. When I remove the redundancy in `addFirstPartyRelationship`, that method only ends up being two lines. To me,  seems like it only saves one line of code; but I can see how it would be useful for clarity.
```javascript
  addFirstPartyRelationship(firstParty, firstPartyHostname, thirdPartyHostname) {
    firstParty['thirdPartyHostnames'].push(target);
    this.setFirstParty(firstPartyHostname, firstParty);
  }
```
4. Also the logic of the nested `if` statements does not account for shared third party sites that are NOT whitelisted for any third party:
Example: visit en.gravatar.com then tylersavin.com. Both link to:
* s0.wp.com
* p.typekit.net
* www.google-analytics.com
* s.gravatar.com (owned by en.gravatar.com)
Because the first 3 third parties are already in firstParty['thirdPartyHostnames'] from visiting en.gravatar.com, their links to this new first party tylersavin.com are not updated.

To fix this, I have created a `shouldUpdate` boolean, which is a combination of three other boolean values:
```javascript
// should update if:
    //   - there is a new third party, OR
    //   - there is a new link between an old third party and a new first party
    //   AND
    //   - the third party is visible
    const shouldUpdate = (isNewThirdParty || isNewFirstParty)
      && thirdParty['isVisible'];

    if (shouldUpdate) {
      this.updateChild(this.outputWebsite(target, thirdParty));
    }
```